src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -221,27 +340,98 @@ static ucs_status_t uct_cuda_copy_mem_alloc(uct_md_h md, size_t *length_p,
         return UCS_ERR_NO_DEVICE;
     }
 
+    alloc_handle = ucs_malloc(sizeof(uct_cuda_copy_alloc_handle_t), "uct_cuda_copy_alloc_handle_t");
+    if (NULL == alloc_handle) {
+        ucs_error("failed to allocate memory for uct_cuda_copy_alloc_handle_t");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    alloc_handle->length = *length_p;
+
     if (mem_type == UCS_MEMORY_TYPE_CUDA) {
-        status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemAlloc((CUdeviceptr*)address_p,
-                                                     *length_p));
+
+        for (method = methods; method < methods + num_methods; ++method) {
+            switch (*method) {
+                case UCT_CUDA_ALLOC_METHOD_FABRIC_PINNED:
+                    status = uct_cuda_copy_mem_alloc_fabric(md, alloc_handle);
+                    if (status != UCS_OK) {
+                        continue;
+                    }
+                    goto allocated;
+                case UCT_CUDA_ALLOC_METHOD_LEGACY_PINNED:
+                    status =
+                        UCT_CUDADRV_FUNC_LOG_ERR(cuMemAlloc(&alloc_handle->ptr,
+                                    alloc_handle->length));
+                    if (status != UCS_OK) {
+                        continue;
+                    }
+                    alloc_handle->alloc_type = *method;
+                    goto allocated;
+                default:
+                    status = UCS_ERR_UNSUPPORTED;
+                    goto err;
+            }
+        }
     } else {
         status =
-            UCT_CUDADRV_FUNC_LOG_ERR(cuMemAllocManaged((CUdeviceptr*)address_p,
-                                                       *length_p,
+            UCT_CUDADRV_FUNC_LOG_ERR(cuMemAllocManaged(&alloc_handle->ptr,
```

Comment:
if we get here , are we sure mem_type is managed? if yes let's add an assertion (or explicit check mem_type==MANAGED)